single_kernel_mongo/managers/sharding.py

```diff
@@ -0,0 +1,689 @@
+    DatabaseRequirerData,
+)
+from single_kernel_mongo.state.charm_state import CharmState
+from single_kernel_mongo.state.config_server_state import SECRETS_FIELDS, ConfigServerKeys
+from single_kernel_mongo.state.tls_state import SECRET_CA_LABEL
+from single_kernel_mongo.utils.mongo_connection import MongoConnection, NotReadyError
+from single_kernel_mongo.utils.mongodb_users import BackupUser, MongoDBUser, OperatorUser
+from single_kernel_mongo.workload.mongodb_workload import MongoDBWorkload
+
+if TYPE_CHECKING:
+    from single_kernel_mongo.managers.mongodb_operator import MongoDBOperator
+
+logger = getLogger(__name__)
+
+
+class ConfigServerManager(Object):
+    """Manage relations between the config server and the shard, on the config-server's side."""
+
+    def __init__(
+        self,
+        dependent: MongoDBOperator,
+        workload: MongoDBWorkload,
+        state: CharmState,
+        substrate: Substrates,
+        relation_name: RelationNames = RelationNames.CONFIG_SERVER,
+    ):
+        super().__init__(parent=dependent, key=relation_name)
+        self.dependent = dependent
+        self.charm = dependent.charm
+        self.state = state
+        self.workload = workload
+        self.substrate = substrate
+        self.relation_name = relation_name
+        self.data_interface = DatabaseProviderData(
+            self.model, relation_name=self.relation_name.value
+        )
+
+    def prepare_sharding_config(self, relation: Relation):
+        """Handles the database requested event.
+
+        It shares the different credentials and necessary files with the shard.
+        """
+        self.assert_pass_hook_checks(relation)
+
+        if self.data_interface.fetch_relation_field(relation.id, "database") is None:
+            raise DeferrableFailedHookChecksError(
+                f"Database Requested event has not run yet for relation {relation.id}"
+            )
+        relation_data = {
+            ConfigServerKeys.operator_password.value: self.state.get_user_password(OperatorUser),
+            ConfigServerKeys.backup_password.value: self.state.get_user_password(BackupUser),
+            ConfigServerKeys.key_file.value: self.state.get_keyfile(),
+            ConfigServerKeys.host.value: json.dumps(sorted(self.state.app_hosts)),
+        }
+
+        if int_tls_ca := self.state.tls.get_secret(internal=True, label_name=SECRET_CA_LABEL):
+            relation_data[ConfigServerKeys.int_ca_secret.value] = int_tls_ca
+
+        self.data_interface.update_relation_data(relation.id, relation_data)
+        self.data_interface.set_credentials(
+            relation.id, "unused", "unused"
+        )  # Triggers the database created event
+
+    def reconcile_shards_for_relation(self, relation: Relation, is_leaving: bool = False):
+        """Handles adding and removing shards.
+
+        Updating of shards is done automatically via MongoDB change-streams.
+        """
+        logger.info("Running Relation Changed hook.")
```

Comment:
maybe this logger could be moved to event handler

Reply:
It even should be removed completely, I just used it as debug log tbh.
I'll remove it